File: lib/models/product_model.dart
// ignore_for_file: public_member_api_docs, sort_constructors_first
import 'dart:convert';

import 'package:amazon_new/models/ratings.dart';

class Product {
  final String name;
  final String description;
  final double quantity;
  final List<String> images;
  final String category;
  final double price;
  final String? id;
  final String? userId;
  final List<Ratings>? ratings;
  Product(
      {required this.name,
      required this.description,
      required this.quantity,
      required this.images,
      required this.category,
      required this.price,
      this.id,
      this.userId,
      this.ratings});

  Map<String, dynamic> toMap() {
    return {
      'name': name,
      'description': description,
      'quantity': quantity,
      'images': images,
      'category': category,
      'price': price,
      'id': id,
      'userId': userId,
      'ratings': ratings,
    };
  }

  factory Product.fromMap(Map<String, dynamic> map) {
    return Product(
        name: map['name'] ?? '',
        description: map['description'] ?? '',
        quantity: map['quantity']?.toDouble() ?? 0.0,
        images: List<String>.from(map['images']),
        category: map['category'] ?? '',
        price: map['price']?.toDouble() ?? 0.0,
        id: map['_id'],
        userId: map['userId'],
        ratings: map['ratings'] != null
            ? List<Ratings>.from(map['ratings'].map((x) => Ratings.fromMap(x)),)
            : null);
  }

  String toJson() => json.encode(toMap());

  factory Product.fromJson(String source) =>
      Product.fromMap(json.decode(source));
}

File: server/models/product_model.js
const express = require("express");
const { Mongoose, default: mongoose } = require("mongoose");
const ratingSchema = require("./ratings");
const ProductSchema = mongoose.Schema({
    name:{
        type:String,
        required:true,
        trim:true
    },
    description:{
        type:String,
        required:true,
        trim:true
    },
    quantity:{
        type:Number,
        required:true
    },
    images:[
        {
            type:String,
            required:true,
        }
    ],
    category:{
        type:String,
        required:true,
    },
    price:{
        type:Number,
        required:true,
    },
    ratings:[ratingSchema],

});

const Product = mongoose.model('Product', ProductSchema);
module.exports = {Product, ProductSchema};